R/log_shiny_input_changes.R

```diff
@@ -37,35 +38,38 @@ log_shiny_input_changes <- function(
     input,
     namespace = NA_character_,
     excluded_inputs = character(),
-    excluded_pattern = "_width$") {
-  session <- shiny::getDefaultReactiveDomain()
-  if (!(shiny::isRunning() || inherits(session, "MockShinySession") || inherits(session, "session_proxy"))) {
-    stop("No Shiny app running, it makes no sense to call this function outside of a Shiny app")
+    excluded_pattern = "_width$",
+    session = shiny::getDefaultReactiveDomain()) {
+  stopifnot(inherits(input, "reactivevalues"))
+  stopifnot(is.character(namespace) && length(namespace) == 1)
+  stopifnot(is.character(excluded_inputs))
+  stopifnot(is.character(excluded_pattern) && length(excluded_pattern) == 1)
+  stopifnot(inherits(session, "session_proxy"))
+
+  if (logger::TRACE > logger::as.loglevel(get_val("TEAL.LOG_LEVEL", "teal.log_level", "INFO"))) {
+    # to avoid setting observers when not needed
+    return(invisible(NULL))
   }
-  ns <- ifelse(!is.null(session), session$ns(character(0)), "")
 
-  # utils::assignInMyNamespace and utils::assignInNamespace are needed
-  # so that observer is executed only once, not twice.
-  input_values <- shiny::isolate(shiny::reactiveValuesToList(input))
-  utils::assignInMyNamespace("shiny_input_values", input_values)
+  ns <- session$ns(character(0))
+  reactive_input_list <- shiny::reactive({
+    input_list <- shiny::reactiveValuesToList(input)
+    input_list[!grepl(excluded_pattern, names(input_list))]
+  })
+  shiny_input_values <- shiny::reactiveVal(shiny::isolate(reactive_input_list()))
 
-  shiny::observe({
-    old_input_values <- shiny_input_values
-    new_input_values <- shiny::reactiveValuesToList(input)
-    names <- unique(c(names(old_input_values), names(new_input_values)))
-    names <- setdiff(names, excluded_inputs)
-    if (length(excluded_pattern)) {
-      names <- grep(excluded_pattern, names, invert = TRUE, value = TRUE)
-    }
+  shiny::observeEvent(reactive_input_list(), {
+    old_input_values <- shiny_input_values()
+    new_input_values <- reactive_input_list()
+    names <- intersect(names(old_input_values), names(new_input_values))
     for (name in names) {
-      old <- old_input_values[name]
-      new <- new_input_values[name]
+      old <- unname(old_input_values[name])
+      new <- unname(new_input_values[name])
       if (!identical(old, new)) {
         message <- trimws("{ns} Shiny input change detected in {name}: {old} -> {new}")
```

Comment:
we must use collapsing mechanism to secure against lenghty `old` and `new` values
```
r$> logger::log_info("Change in {name}: {1:3} -> {1:2}")
Error in value[[3L]](cond) : `glue` failed in `formatter_glue` on:

  chr "Change in {name}: {1:3} -> {1:2}" 

Raw error message:

 Variables must be length 1 or 3 

Please consider using another `log_formatter` or `skip_formatter` on strings with curly braces.

r$> logger::log_info("Change in {name}: {glue::glue_collapse(1:3, sep = \", \")} -
    > {glue::glue_collapse(1:2, sep = \", \")}")
INFO [2024-09-05 12:15:51] Change in x: 1, 2, 3 -> 1, 2
```

PS. How this hasn't been identified already?

Reply:
> PS. How this hasn't been identified already?

It hasn't been detected because it is not the case at this moment. `old` and `new` are always a list. Only if you change the `glue_formatter` to the one you proposed [here](https://github.com/insightsengineering/teal.logger/pull/86#discussion_r1745151365) then this problem is introduced. I think keeping `old` new as `list` is the simplest solution to handle:

- this
```r
vec1 <- list(c("a", "b"))
vec2 <- list(c("a", "b", "c"))
glue::glue("{vec1} -> {vec2}")
# c("a", "b") -> c("a", "b", "c")
```

- and this
```r
vec1 <- list(NULL)
vec2 <- list(c("a", "b", "c"))
glue::glue("{vec1} -> {vec2}")
# NULL -> c("a", "b", "c")
```

Reply:
OK thanks. That makes sense.

Re: `old` and `new` as list - I'm still not super convinced to be honest. It's not very clear that keeping this as a list determines correct formatting (of multiple values and NULL and possibly other things). This should be handled in the _formatting part_ of this is is a formatter component to the logger. That's why I think we should apply custom glue formatter with collapse.

Reply:
I guess we would need to add a `.transformer` to handle atomic vectors to `glue_formatter` [See this](https://glue.tidyverse.org/articles/transformers.html#variables-and-values-transformer). Then 

```r
teal_logger_transformer <- function(text, envir) {
  value <- glue::identity_transformer(text, envir)
  if (is.null(value)) {
    "NULL"
  } else if (is.atomic(value) && length(value) == 0) {
    sprintf("%s(0)", typeof(value))
  } else if (is.atomic(value) && length(value) > 1) {
    sprintf("[%s]", toString(value))
  } else {
    glue::glue_collapse(value, sep = ", ")
  }
}

logger::log_formatter(
  function(..., .logcall = sys.call(), .topcall = sys.call(-1), .topenv = parent.frame()) {
    logger::formatter_glue(
      ..., .null = "NULL", .logcall = .logcall, .topcall = .topcall, .topenv = .topenv,
      .transformer = teal_logger_transformer
    )
  }
)

logger::log_info("list(6): {list(letters[1:6])} and list(3): {list(letters[1:3])}.")
# INFO [2024-09-05 16:33:58] list(6): c("a", "b", "c", "d", "e", "f") and list(3): c("a", "b", "c").

logger::log_info("character(6): {letters[1:6]} and character(3) {letters[1:3]}.")
# INFO [2024-09-05 16:33:59] character(6): [a, b, c, d, e, f] and character(3) [a, b, c].

logger::log_info("list(6): {list(letters[1:6])} and character(3): {letters[1:3]}.")
# INFO [2024-09-05 16:33:59] list(6): c("a", "b", "c", "d", "e", "f") and character(3): [a, b, c].

logger::log_info("nested list: {list(a = list(b = letters[1:6]), letters[7:8])}.")
# INFO [2024-09-05 16:33:59] nested list: list(b = c("a", "b", "c", "d", "e", "f")), c("g", "h").

logger::log_info("NULL value: {NULL}")
# INFO [2024-09-05 16:34:00] NULL value: NULL

logger::log_info("NULL value: {character(0)}")
# INFO [2024-09-05 16:34:00] NULL value: character(0)

logger::log_info("NULL value: {integer(0)}")
# INFO [2024-09-05 16:34:00] NULL value: integer(0)

logger::log_info("NULL value: {NA}")
# INFO [2024-09-05 16:34:02] NULL value: NA
```

Reply:
custom transformer had been pushed to the codebase. closing this thread